Microsoft.NET.Build.Containers/Tasks/CreateNewImage.cs

```diff
@@ -18,24 +18,115 @@ public partial class CreateNewImage : Microsoft.Build.Utilities.Task
+        if (Log.HasLoggedErrors)
+        {
+            return false;
+        }
+
+        // at this point we're done with modifications and are just pushing the data other places
+        GeneratedContainerManifest = JsonSerializer.Serialize(image.Manifest);
+        GeneratedContainerConfiguration = image.Config.ToJsonString();
+
+        foreach (var destinationImageReference in destinationImageReferences)
+        {
+            if (IsDaemonPush)
+            {
+                var localDaemon = GetLocalDaemon(msg => Log.LogMessage(msg));
+                if (!localDaemon.IsAvailable().GetAwaiter().GetResult())
+                {
+                    Log.LogError("The local daemon is not available, but pushing to a local daemon was requested. Please start the daemon and try again.");
+                    return false;
+                }
+                try
+                {
+                    localDaemon.Load(image, sourceImageReference, destinationImageReference).Wait();
+                    SafeLog("Pushed container '{0}' to local daemon", destinationImageReference.RepositoryAndTag);
+                }
+                catch (AggregateException ex) when (ex.InnerException is DockerLoadException dle)
+                {
+                    Log.LogErrorFromException(dle, showStackTrace: false);
+                }
+            }
+            else
+            {
+                try
+                {
+                    destinationImageReference.Registry?.Push(image, sourceImageReference, destinationImageReference, message => SafeLog(message)).Wait();
+                    SafeLog("Pushed container '{0}' to registry '{2}'", destinationImageReference.RepositoryAndTag, OutputRegistry);
+                }
+                catch (ContainerHttpException e)
+                {
+                    if (BuildEngine != null)
+                    {
+                        Log.LogErrorFromException(e, true);
+                    }
+                }
+                catch (Exception e)
+                {
+                    if (BuildEngine != null)
+                    {
+                        Log.LogError("Failed to push to the output registry: {0}", e);
+                    }
+                }
+            }
+        }
+
+        return !Log.HasLoggedErrors;
+    }
 
     private void SetPorts(Image image, ITaskItem[] exposedPorts)
     {
         foreach (var port in exposedPorts)
         {
             var portNo = port.ItemSpec;
-            var portTy = port.GetMetadata("Type");
-            if (ContainerHelpers.TryParsePort(portNo, portTy, out var parsedPort, out var errors))
+            var portType = port.GetMetadata("Type");
+            if (ContainerHelpers.TryParsePort(portNo, portType, out Port? parsedPort, out ContainerHelpers.ParsePortError? errors))
             {
-                image.ExposePort(parsedPort.number, parsedPort.type);
+                image.ExposePort(parsedPort.Value.Number, parsedPort.Value.Type);
```

Comment:
I missed why the additional calls to `.Value` started happening for port - can you help me understand why that changed?

Reply:
Since `Port` is now `record struct`, and the param is nullable we have to use `Value`. Removing nullable or keeping it as class (or record) might be an option too. `record struct` looks better due to implicit `IEquatable` implementation.